projects/angular/src/forms/file-input/file-input-container.ts

```diff
@@ -75,28 +97,84 @@ import { ClrFileInput } from './file-input';
 
-  protected readonly commonStrings = inject(ClrCommonStringsService);
-
-  @ContentChild(forwardRef(() => ClrFileInput)) protected readonly fileInput: ClrFileInput;
+  @ContentChild(forwardRef(() => ClrFileInput)) readonly fileInput: ClrFileInput;
+  @ContentChild(forwardRef(() => ClrFileList)) protected readonly fileList: ClrFileList;
 
   @ViewChild('browseButton') private browseButtonElementRef: ElementRef<HTMLButtonElement>;
+  @ViewChild('fileListFileInput') private fileListFileInputElementRef: ElementRef<HTMLInputElement>;
+
+  // These are for the "message present" override properties
+  @ContentChild(ClrFileSuccess) private readonly fileSuccessComponent: ClrFileSuccess;
+  @ContentChild(ClrFileError) private readonly fileErrorComponent: ClrFileError;
+
+  private readonly commonStrings = inject(ClrCommonStringsService);
+
+  protected get accept() {
+    return this.fileInput.elementRef.nativeElement.accept;
+  }
+
+  protected get multiple() {
+    return this.fileInput.elementRef.nativeElement.multiple;
+  }
 
   protected get disabled() {
     return this.fileInput.elementRef.nativeElement.disabled;
   }
 
+  protected get browseButtonText() {
+    const selectionButtonLabel = this.fileList ? undefined : this.fileInput?.selection?.buttonLabel;
+
+    return selectionButtonLabel || this.customButtonLabel || this.commonStrings.keys.browse;
+  }
+
   protected get browseButtonDescribedBy() {
-    return `${this.label?.idAttr} ${this.fileInput.elementRef.nativeElement.getAttribute('aria-describedby')}`;
+    return `${this.label?.forAttr} ${this.fileInput.elementRef.nativeElement.getAttribute('aria-describedby')}`;
+  }
+
+  protected override get successMessagePresent() {
+    return super.successMessagePresent || !!this.fileSuccessComponent;
+  }
+
+  protected override get errorMessagePresent() {
+    return super.errorMessagePresent || !!this.fileErrorComponent;
+  }
+
+  focusBrowseButton() {
+    this.browseButtonElementRef.nativeElement.focus();
   }
 
   protected browse() {
-    this.fileInput.elementRef.nativeElement.click();
+    const fileInputElementRef =
+      this.fileList && this.multiple ? this.fileListFileInputElementRef : this.fileInput.elementRef;
+
+    fileInputElementRef.nativeElement.click();
   }
 
   protected clearSelectedFiles() {
     this.fileInput.elementRef.nativeElement.value = '';
     this.fileInput.elementRef.nativeElement.dispatchEvent(new Event('change', { bubbles: true, cancelable: true }));
 
-    this.browseButtonElementRef.nativeElement.focus();
+    this.focusBrowseButton();
+  }
+
+  protected addFilesToSelection(newFiles: FileList) {
+    if (!newFiles.length) {
+      return;
+    }
+
+    const existingFiles = Array.from(this.fileInput.elementRef.nativeElement.files);
+
+    // start with new files
+    const mergedFiles = [...Array.from(newFiles)];
+
+    // add existing files if a new file doesn't have the same name
+    for (const existingFile of existingFiles) {
```

Comment:
Move `existingFiles` creation closer to the loop for better visiblity.
```suggestion
    // add existing files if a new file doesn't have the same name
    const existingFiles = this.fileInput.elementRef.nativeElement.files;
    for (const existingFile of existingFiles) {
```

Technically can remove the local `const` altogether.
```suggestion
    // add existing files if a new file doesn't have the same name
    for (const existingFile of this.fileInput.elementRef.nativeElement.files) {
```

Reply:
Yeah, I think I agree. I'll simplify this. Thanks!